Library/Homebrew/utils/ast.rb

```diff
@@ -114,6 +114,39 @@ def add_bottle_block(bottle_output)
         add_stanza(:bottle, "\n#{bottle_output.chomp}", type: :block_call)
       end
 
+      sig { params(name: Symbol, type: T.nilable(Symbol)).void }
+      def remove_stanza(name, type: nil)
+        stanza_node = stanza(name, type: type)
+        raise "Could not find '#{name}' stanza!" if stanza_node.blank?
+
+        # stanza is probably followed by a newline character
+        # try to delete it if so
+        stanza_range = stanza_node.source_range
+        trailing_range = stanza_range.with(begin_pos: stanza_range.end_pos,
+                                           end_pos:   stanza_range.end_pos + 1)
+        if trailing_range.source.chomp.empty?
+          stanza_range = stanza_range.adjust(end_pos: 1)
+
+          # stanza_node is probably indented
+          # since a trailing newline has been removed,
+          # try to delete leading whitespace on line
+          leading_range = stanza_range.with(begin_pos: stanza_range.begin_pos - stanza_range.column,
+                                            end_pos:   stanza_range.begin_pos)
+          if leading_range.source.strip.empty?
+            stanza_range = stanza_range.adjust(begin_pos: -stanza_range.column)
+
+            # if the stanza was preceded by a blank line, it should be removed
+            # that is, if the two previous characters are newlines,
+            # then delete one of them
```

Comment:
One thought about this: a formula file that has newlines separated by whitespace is still valid. For example, if the `.` in the following formula represents a space, the following will still be valid:

```ruby
class Hello < Formula
  desc "foo"
..
  bottle do
    sha256 "f7b1fc772c79c20fddf621ccc791090bc1085fcef4da6cca03399424c66e06ca" => :sierra
  end
end
```

I do have a style rule that ensures that formulae don't have trailing whitespace so we should be totally fine in homebrew/core, but it might be worth noting that this isn't necessarily true for other taps.

Reply:
I think it won't strip whitespace in that case, though it will leave an extra blank line. I think it would look like the following after a `remove_stanza(:bottle)` call (with `.` representing a space):

~~~
class Hello < Formula
  desc "foo"
..

end
~~~

My goal isn't to handle all whitespace edge cases, but rather to avoid introducing style errors to clean files. I can try to improve the documentation if it's not currently clear enough.

Reply:
No, I think it's totally fine as is. Agreed, as long as it handles the standard cases it's fine. Plus we have `brew style --fix` to help out in other situations too.